consensus/src/lib.rs

```diff
@@ -1,27 +1,1037 @@
+                if block_source != BlockSource::Local {
+                    return Err(BlockError::InvalidBlockSource);
+                };
+            }
+        }
+
+        self.check_transactions(block)?;
+        Ok(())
+    }
+
+    fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
+        let _consensus_data = block.consensus_data();
+        // TODO: PoW is not in master at the moment =(
+        Ok(())
+    }
+
+    fn check_transactions(&self, block: &Block) -> Result<(), BlockError> {
+        // TODO: Must check for duplicate inputs (see CVE-2018-17144)
+        //      We should discuss - can we add Hash trait to Transaction?
+        //      We will have plenty more checks with inputs\outputs and HashSet\BTreeMap might be more efficient
+        //
+        // let mut keyed = HashSet::new();
+        // for tx in block.get_transactions() {
+        //     for input in tx.get_inputs() {
+        //         if keyed.contains(input.get_outpoint()) {
+        //             return Err(BlockError::Unknown);
+        //         }
+        //         keyed.insert(input.get_outpoint());
+        //     }
+        // }
+
+        //TODO: Size limits
+        if block.encoded_size() > MAX_BLOCK_WEIGHT {
+            return Err(BlockError::Unknown);
+        }
+        //TODO: Check signatures will be added when will ready BLS
+        Ok(())
+    }
+
+    fn check_block(&self, block: &Block, block_source: BlockSource) -> Result<(), BlockError> {
+        self.check_consensus(block)?;
+        self.check_block_detail(block, block_source)?;
+        Ok(())
+    }
+
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> bool {
+        block_index.get_next_block_id().is_some()
+            || self
+                .db_tx
+                .get_best_block_id()
+                .ok()
+                .flatten()
+                .map_or(false, |ref block_id| block_index.get_block_id() == block_id)
+    }
+
+    /// Mark new block as an orphan
+    fn new_orphan_block(&mut self, block: Block) -> Result<(), BlockError> {
+        // It can't be a genesis block
+        assert!(!block.is_genesis(self.chain_config));
+        self.orphan_blocks.add_block(block).map_err(|err| match err {
+            OrphanAddError::BlockAlreadyInOrphanList(_) => BlockError::Orphan,
+        })?;
+        Ok(())
+    }
+
+    fn get_block_from_index(&self, block_index: &BlockIndex) -> Result<Option<Block>, BlockError> {
+        Ok(self.db_tx.get_block(block_index.get_block_id().clone())?)
+    }
+}
 
 #[cfg(test)]
 mod tests {
+    use super::*;
+    use blockchain_storage::Store;
+    use common::address::Address;
+    use common::chain::block::{Block, ConsensusData};
+    use common::chain::config::create_mainnet;
+    use common::chain::{Destination, Transaction, TxInput, TxOutput};
+    use common::primitives::{Amount, Id};
+
+    fn produce_test_block(config: &ChainConfig, prev_block: &Block, orphan: bool) -> Block {
```

Comment:
You either know the previous block or it's an orphan. These are mutually exclusive. Why two parameters, and not one enum with both options?